src/pointarray.cpp

```diff
@@ -387,10 +506,53 @@ V3d PointArray::pickVertex(const V3d& cameraPos,
                            double longitudinalScale, double* distance,
                            std::string* info) const
 {
+    // NOTE rayOrigin here is not cameraPos but something else
+    // This routine uses cameraPos as the rayOrigin point
+
     if (m_npoints == 0)
         return V3d(0);
-    size_t idx = closestPointToRay(m_P, m_npoints, rayOrigin - offset(),
-                                   rayDirection, longitudinalScale, distance);
+
+    const V3f cameraPosOffset = cameraPos - offset();
+
+    float rToAxis    = FLT_MAX;
+    float xToOrigin  = FLT_MAX;
+    size_t iToAxis   = 0;
+    size_t iToOrigin = 0;
+
+    std::vector<const OctreeNode*> nodeStack;
+    nodeStack.push_back(m_rootNode.get());
+    while (!nodeStack.empty())
+    {
+        const OctreeNode* node = nodeStack.back();
+        nodeStack.pop_back();
+
+        if (!node->isLeaf())
+        {
+            for (int i = 0; i < 8; ++i)
+            {
+                OctreeNode* n = node->children[i];
+                if (n)
+                    nodeStack.push_back(n);
+            }
+            continue;
+        }
+
+        node->closestPoint(m_P, cameraPosOffset, rayDirection, rToAxis, xToOrigin, iToAxis, iToOrigin);
+    }
+
+    size_t idx;
+
+    if(xToOrigin == FLT_MAX)
+    {
+        // no points within cylinder
+        if(rToAxis == FLT_MAX)
+            idx = m_lastVertexPickIdx; // no nearest point to axis found either
```

Comment:
I don't think having `m_lastVertexPickIdx` here really makes sense - it's papering over the fact that nothing valid was found, and just returning something anyway.

After some thought, I'd say there's a couple of viable alternatives:
* Change the semantics of `pickVertex()` so that there's a way of signaling that no valid vertex was found.  The calling code then has to deal with it.  (Does it make sense to return a point when all vertices are behind the camera?  Kinda not, but it does allow the user to bring the data into view regardless of how they've messed up their camera.) 
* Change the code above so that the tree traversal happens in order of octree node proximity to the ray, and proceeds until it can be proven that the remaining nodes are further away than the currently selected closest point.  I favour doing this regardless, since it allows some neat things with selection painting which I can show tomorrow.  It should allow the previous selection distance algorithm to be accelerated in the octree framework.